lib/Subscription/Manage.php

```diff
@@ -48,6 +54,19 @@ function onSave() {
     $this->url_helper->redirectBack();
   }
 
+  private function addHiddenSegments(Subscriber $subscriber, array $data) {
+    $shown_segments = $this->settings->get('subscription.segments');
+    if (empty($shown_segments)) return $data;
+    if (empty($data['segments'])) $data['segments'] = [];
+    $subscriber->withSubscriptions();
+    foreach ($subscriber->subscriptions as $subscription) {
+      if ($subscription['status'] === Subscriber::STATUS_SUBSCRIBED && !in_array($subscription['segment_id'], $shown_segments)) {
+        $data['segments'][] = $subscription['segment_id'];
+      }
+    }
```

Comment:
This is quite a hard to read logic and there's no comment to explain it. But I don't like it anyway - it seems we take data from user's form and then add some stuff to it from the DB (and only specific subscription status). Why don't we do naturally, the other way round? I think we should first fetch all subscription data from the DB, then apply user changes on it, then save it. That would be a more straightforward and safe way to do it.